fhir2dataset/fhirrules_getter.py

```diff
@@ -43,74 +44,42 @@ def __init__(
-            name_search_param = resource_tree.execute("$.resource.code")
-            expression_search_param = resource_tree.execute("$.resource.expression")
-            # xpath_search_param = resource_tree.execute("$.resource.xpath")
-            for resource_type in resource_tree.execute("$.resource.base"):
-                if resource_type not in dict_searchparam:
-                    dict_searchparam[resource_type] = dict()
-                if expression_search_param:
-                    list_expression = expression_search_param.split(" | ")
-                    find = False
-                    for exp in list_expression:
-                        exp_split = exp.split(".")
-                        if resource_type == exp_split[0]:
-                            find = True
-                            expression = ".".join(exp_split[1:])
-                        elif resource_type in exp_split[0]:
-                            # special cases : CodeableConcept, Quantity etc
-                            find = True
-                            exp_split = ".".join(exp_split[1:])
-                            expression = exp_split.split(" ")[0]
-                            logger.debug(
-                                f"\nthe searchpram '{name_search_param} is associated with this"
-                                f" FHIRpath '{expression}'\n"
-                            )
-                    if not find:
-                        logger.warning(
-                            f"\nthe fhirpath associated to the instance of SearchParamater "
-                            f"named '{name_search_param}' wasn't found for the resource_type "
-                            f"'{resource_type}''"
-                        )
-                        logger.debug(
-                            f"the list of firpath associated with this SearchParamater is "
-                            f"{list_expression}"
-                            f"the json associated is :\n {resource}\n"
-                        )
-                    dict_searchparam[resource_type][name_search_param] = {
-                        "expression": expression,
-                    }
-                else:
-                    logger.warning(
-                        f"\nthe instance of SearchParamater named '{name_search_param}'"
-                        f" has no expression associated"
-                    )
-                    logger.debug(f"{resource}\n")
+        dict_searchparam = defaultdict(dict)
+        fhirpaths = ["SearchParameter.code", "SearchParameter.expression", "SearchParameter.base"]
+        resources = [resource["resource"] for resource in self.searchparameters["entry"]]
+        results = multiple_search_dict(resources, fhirpaths)
+        for idx, result in enumerate(results):
+            if result[1]:
+                for resource_type in result[2]:
+                    dict_searchparam[resource_type][result[0][0]] = result[1][0]
```

Comment:
The structure of `result` is a bit obscur to me, maybe we could use a class here or smthg more explicit?

Reply:
Indeed it is not clear. Does the introduction of the `dataclasses elements`, the modification of the `multiple_search_dict` function and the `MAPPING_SEARCHPARAMS` mapping make the code more readable here?